innative/atomic_instructions.cpp

```diff
@@ -0,0 +1,256 @@
+// Copyright (c)2020 Black Sphere Studios
+// For conditions of distribution and use, see copyright notice in innative.h
+
+#include "llvm.h"
+#include "compile.h"
+#include "utility.h"
+#include "validate.h"
+#include "atomic_instruction_details.h"
+
+using namespace innative;
+using namespace utility;
+using Func    = llvm::Function;
+using FuncTy  = llvm::FunctionType;
+using llvmTy  = llvm::Type;
+using llvmVal = llvm::Value;
+
+IN_ERROR innative::Compiler::InsertAlignmentTrap(llvmVal* ptr, varuint32 memory, varuint32 memflags)
+{
+  if(!memflags) // Only bother if alignment > 1
+    return ERR_SUCCESS;
+
+  auto alignment = 1Ui64 << memflags;
+  auto sizet     = builder.getIntNTy(machine->getPointerSizeInBits(memory));
+  auto ptrtoint  = builder.CreatePtrToInt(ptr, sizet);
+  auto zero      = CInt::get(sizet, 0);
+  auto mask      = CInt::get(sizet, alignment - 1);
+  auto cond      = builder.CreateICmpNE(builder.CreateAnd(ptrtoint, mask), zero);
+  InsertConditionalTrap(cond);
+
+  return ERR_SUCCESS;
+}
+
+IN_ERROR innative::Compiler::InsertAtomicMemGet(Instruction& ins, llvmVal*& ptr, unsigned& align)
+{
+  IN_ERROR err;
+
+  llvmVal* base;
+  if(err = PopType(TE_i32, base))
+    return err;
+
+  varuint32 memflags = ins.immediates[0]._varuint32;
+  varuint32 offset   = ins.immediates[1]._varuint32;
+  varuint32 memory   = ins.immediates[2]._varuint32;
```

Comment:
I am unable to verify in the spec that this is the expected type of the memory index. Do you have a link or an example of a test that shows it must be considered a 32-bit integer?